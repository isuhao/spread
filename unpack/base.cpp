#include "base.hpp"

#include "dirwriter.hpp"

using namespace Mangle::VFS;
using namespace Mangle::Stream;
using namespace Unpack;

// "Convert" directory name to Mangle::StreamFactory
void UnpackBase::unpack(const std::string &file,
                        const std::string &dir,
                        Progress *prog,
                        const FileList &list)
{
  StreamFactoryPtr ptr(new Unpack::DirWriter(dir));
  unpack(file, ptr, prog, list);
}

struct ListMaker : StreamFactory
{
  UnpackBase::FileList *list;
  StreamFactoryPtr redir;

  StreamPtr open(const std::string &name)
  {
    list->insert(name);
    if(redir)
      return redir->open(name);
    return StreamPtr();
  }
};

void UnpackBase::index(const std::string &file,
                       FileList &output,
                       Progress *prog,
                       StreamFactoryPtr write)
{
  ListMaker *m = new ListMaker;
  m->list = &output;
  m->redir = write;
  StreamFactoryPtr mp(m);

  // This basically unpacks the file normally, and pumps all the data
  // through ListMaker. The ListMaker makes note of all the filenames
  // given, and store them in 'output'.
  unpack(file, mp, prog);
}
