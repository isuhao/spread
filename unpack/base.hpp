#ifndef __UNPACK_BASE_HPP
#define __UNPACK_BASE_HPP

#include <mangle/vfs/stream_factory.hpp>
#include <set>

#include "progress.hpp"

namespace Unpack
{
  /*
    How the VFS::StreamFactory is used:

    - open("filename") is called for all files in the archive. The
      returned stream is expected to be writable, except for directory
      entries. The data for each file is written to the stream, and
      the StreamPtr is dereferenced.

    - Returning an empty StreamPtr is allowed, and will cause the file
      to be skipped. This can be used to index the archive.

    - Filenames may include paths. The reciever is expected to create
      necessary parent directories.

    - Directories (paths ending with a slash) may be sent to
      open(). The StreamFactory is expected to create empty
      directories in those cases, and to return an empty StreamPtr.
      Both forward and backward slashes should be expected.

    - The StreamFactory itself is responsible for throwing exceptions
      on writing errors. The unpacker will only throw on unpacking
      errors.

    - Instead of using StreamFactory directly, you may use the
      directory version to write to a file system directory.

    About the FileList list:

    - If the 'list' parameter is specified, only the files listed
      there will be unpacked. Files in the list which do not match any
      in the archive, will cause an error.

      Names must match exactly with the files in the archive,
      including case and slash type. For that reason, it is recommened
      to ONLY use filenames that have previously been generated with
      the same unpacker.

    - If the list is missing or empty, all files are unpacked.

    - Some implementations may behave differently depending on whether
      the list is present or not. Some EXE unpackers for example will
      enter a slow "explore mode" when a list is missing, but will be
      in a much faster "unpack mode" when given the right data. In
      those cases, the list names may contain additional meta-data
      that do not really represent actual filenames.
   */
  struct UnpackBase
  {
    typedef std::set<std::string> FileList;

    // Unpack the given archive file into the stream factory.
    virtual void unpack(const std::string &file,
                        Mangle::VFS::StreamFactoryPtr output,
                        Progress *prog = NULL,
                        const FileList &list = FileList()) = 0;

    // Directory writer version
    void unpack(const std::string &file, const std::string &dir,
                Progress *prog = NULL,
                const FileList &list = FileList());

    /* Generate an index of an archive. This is the same as running
       unpack() and gathering up the names generated through the
       StreamFactory.

       An optional StreamFactoryPtr may be provided to accept output
       data. If it is empty, no data is extracted.
    */
    void index(const std::string &file, FileList &output,
               Progress *prog = NULL,
               Mangle::VFS::StreamFactoryPtr write =
               Mangle::VFS::StreamFactoryPtr());
  };
}
#endif
