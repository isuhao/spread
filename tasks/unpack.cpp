#include "unpack.hpp"
#include <unpack/auto.hpp>

using namespace Tasks;

struct UProgress : Unpack::Progress
{
  Jobify::JobInfoPtr info;

  bool progress(int64_t total, int64_t now)
  {
    info->current = now;
    info->total = total;

    // Abort the download if the user requested it.
    if(info->checkStatus())
      return false;

    return true;
  }
};

void UnpackTask::doJob()
{
  UProgress prog;
  prog.info = info;

  Unpack::AutoUnpack unp;

  if(dir != "")
    {
      setBusy("Unpacking " + file + " to " + dir);
      unp.unpack(file, dir, &prog, list);
    }
  else
    {
      setBusy("Unpacking " + file);
      unp.unpack(file, writeTo, &prog, list);
    }

  // All error handling is done through exceptions. If we get here,
  // everything is OK.
  setDone();
};
