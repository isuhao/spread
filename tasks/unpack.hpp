#ifndef __TASKS_UNPACK_HPP_
#define __TASKS_UNPACK_HPP_

#include <job/job.hpp>
#include <mangle/vfs/stream_factory.hpp>
#include <set>

/*
  This class takes any compressed archive file and unpacks it into the
  given stream factory or file system directory.

  The file may be in any known archive format, and the format is
  auto-detected. The backend code is found in the unpack/ module.
 */

namespace Tasks
{
  struct UnpackTask : Jobify::Job
  {
    typedef std::set<std::string> FileList;

    UnpackTask(const std::string &_file, const std::string &_dir,
               Jobify::JobInfoPtr info = Jobify::JobInfoPtr(),
               const FileList &_list = FileList())
      : Jobify::Job(info), file(_file), dir(_dir), list(_list) {}

    UnpackTask(const std::string &_file, Mangle::VFS::StreamFactoryPtr _writeTo,
               Jobify::JobInfoPtr info = Jobify::JobInfoPtr(),
               const FileList &_list = FileList())
      : Jobify::Job(info), file(_file), writeTo(_writeTo), list(_list) {}

  private:
    void doJob();

    std::string file, dir;
    Mangle::VFS::StreamFactoryPtr writeTo;
    const FileList &list;
  };
}

#endif
