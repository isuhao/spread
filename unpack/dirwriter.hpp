#ifndef __UNPACK_DIRWRITER_HPP
#define __UNPACK_DIRWRITER_HPP

#include <mangle/vfs/stream_factory.hpp>

namespace Unpack
{
  /* A stream factory that writes files to a file system directory.
     Directories will be created as necessary. Fully implements all
     UnpackBase necessities.
  */
  struct DirWriter : Mangle::VFS::StreamFactory
  {
    DirWriter(const std::string &dir)
      : base(dir) {}

    Mangle::Stream::StreamPtr open(const std::string &name);

    std::string base;
  };
}

#endif
