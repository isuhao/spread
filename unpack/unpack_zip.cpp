#include "unpack_zip.hpp"

#include <zzip/zzip.h>
#include <stdexcept>
#include <vector>

using namespace Unpack;
using namespace Mangle::Stream;

static void fail(const std::string &msg, const std::string &file)
{
  throw std::runtime_error("Error unpacking "+file+": " + msg);
}

static void checkErr(int err, const std::string &file)
{
  if(err != ZZIP_NO_ERROR)
    fail(zzip_strerror(err), file);
}

void UnpackZip::unpack(const std::string &file, Mangle::VFS::StreamFactoryPtr output,
                       Progress *prog, const FileList &list)
{
  assert(output);

  zzip_error_t err;
  ZZIP_DIR *root = zzip_dir_open(file.c_str(), &err);
  checkErr(err, file);

  // Build a directory of all the files in the archive, and count up
  // the total size
  std::vector<std::string> dir;
  int64_t total = 0, current = 0;
  {
    ZZIP_DIRENT ent;
    while(zzip_dir_read(root, &ent))
      {
        std::string name(ent.d_name);

        if(list.size())
          // Is this file on the extract list?
          if(list.count(name) == 0)
            // Nope. Skip it.
            continue;

        dir.push_back(name);
        total += ent.st_size;
      }
    checkErr(zzip_error(root), file);
  }

  if(list.size() > dir.size())
    fail("Missing files in archive", file);

  bool abort = false;

  // Update progress and check for abort status
  if(prog)
    abort = !prog->progress(total, current);

  for(int i=0; i<dir.size(); i++)
    {
      if(abort)
        break;

      std::string fname = dir[i];

      // Fetch a writable stream
      StreamPtr outs = output->open(fname);
      if(!outs) continue;
      assert(outs->isWritable);

      // Open the archive entry
      ZZIP_FILE *zf = zzip_file_open(root, fname.c_str(), 0);
      if(!zf)
        {
          checkErr(zzip_error(root), file);
          fail("Unknown ZIP error", file);
        }

      while(!abort)
        {
          char buf[10*1024];
          int r = zzip_file_read(zf, buf, 10*1024);

          if(r<0)
            checkErr(zzip_error(root), file);

          outs->write(buf, r);

          current += r;

          // Update progress and check for abort status
          if(prog)
            abort = !prog->progress(total, current);

          if(r < 1024)
            break;
        }
      zzip_file_close(zf);
    }

  zzip_dir_close(root);
}
