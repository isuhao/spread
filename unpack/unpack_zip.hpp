#ifndef __UNPACK_ZIP_HPP
#define __UNPACK_ZIP_HPP

#include "base.hpp"

namespace Unpack
{
  /* ZIP file unpacker.

     This implementation uses the ZZIP library.
   */
  struct UnpackZip : UnpackBase
  {
    void unpack(const std::string &file,
                Mangle::VFS::StreamFactoryPtr output,
                Progress *prog = NULL,
                const FileList &list = FileList());

    using UnpackBase::unpack;
  };
}
#endif
