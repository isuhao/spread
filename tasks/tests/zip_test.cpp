#include "../unpack.hpp"

#include <iostream>
using namespace std;

int main()
{
  string zip = "../../unpack/tests/archives/test.zip";
  cout << "Unpacking " << zip << "...\n";

  Tasks::UnpackTask unp(zip, "_outdir1");
  Jobify::JobInfoPtr info = unp.getInfo();

  unp.run();

  if(info->isSuccess())
    cout << "Success!\n";
  else
    cout << "Failure: " << info->message << endl;

  cout << "Progress: " << info->current << "/" << info->total << endl;

  return 0;
}
