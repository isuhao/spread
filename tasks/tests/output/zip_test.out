Unpacking ../../unpack/tests/archives/test.zip...
Success!
Progress: 491/491
